### ISSUE TITLE
⚙️ Implement Optimization Algorithms

### ISSUE THREAD

**Issue by sakshambedi** [sakshambedi/micrograd#20]
# ⚙️ Implement Optimization Algorithms

## 📋 Description

Implement essential optimization algorithms for training neural networks and machine learning models. Optimizers are crucial for updating model parameters based on computed gradients during the training process.

## 🎯 Acceptance Criteria

- [ ] Implement `SGD` (Stochastic Gradient Descent) with momentum support
- [ ] Implement `Adam` optimizer with adaptive learning rates
- [ ] Implement `RMSprop` optimizer for non-stationary objectives
- [ ] Add `AdamW` with weight decay decoupling
- [ ] Implement learning rate scheduling support
- [ ] Add parameter grouping for different learning rates
- [ ] Support for gradient clipping
- [ ] Implement `zero_grad()` functionality across all optimizers
- [ ] Add state management for optimizer internal variables
- [ ] Support for different parameter types and devices

### Current State
- Optimizers: ❌ Not implemented
- Parameter management: ✅ Parameter class exists (Issue #08)
- Gradient computation: ✅ Available through autograd
- Learning rate scheduling: ❌ Not implemented

### Implementation Requirements

**Base Optimizer Class**
```python
class Optimizer:
    """Base class for all optimizers."""
    
    def __init__(self, params, defaults):
        self.defaults = defaults
        self.state = {}
        self.param_groups = []
        
        if isinstance(params, (list, tuple)):
            params = list(params)
        else:
            params = list(params)
        
        if len(params) == 0:
            raise ValueError("optimizer got an empty parameter list")
        
        param_groups = [{'params': params}]
        
        for param_group in param_groups:
            self.add_param_group(param_group)
    
    def add_param_group(self, param_group):
        """Add a param group to the optimizer."""
        for key, value in self.defaults.items():
            param_group.setdefault(key, value)
        self.param_groups.append(param_group)
    
    def zero_grad(self):
        """Zero gradients for all parameters."""
        for group in self.param_groups:
            for p in group['params']:
                if p.grad is not None:
                    p.grad = None
    
    def step(self):
        """Perform a single optimization step."""
        raise NotImplementedError
```

**SGD Optimizer**
```python
class SGD(Optimizer):
    """Stochastic Gradient Descent optimizer."""
    
    def __init__(self, params, lr=1e-3, momentum=0, dampening=0, 
                 weight_decay=0, nesterov=False):
        if lr < 0.0:
            raise ValueError(f"Invalid learning rate: {lr}")
        if momentum < 0.0:
            raise ValueError(f"Invalid momentum value: {momentum}")
        if weight_decay < 0.0:
            raise ValueError(f"Invalid weight_decay value: {weight_decay}")
        
        defaults = dict(lr=lr, momentum=momentum, dampening=dampening,
                       weight_decay=weight_decay, nesterov=nesterov)
        super().__init__(params, defaults)
    
    def step(self):
        """Perform a single optimization step."""
        for group in self.param_groups:
            weight_decay = group['weight_decay']
            momentum = group['momentum']
            dampening = group['dampening']
            nesterov = group['nesterov']
            
            for p in group['params']:
                if p.grad is None:
                    continue
                
                grad = p.grad
                
                # Apply weight decay
                if weight_decay != 0:
                    grad = grad + weight_decay * p
                
                # Apply momentum
                if momentum != 0:
                    param_state = self.state.setdefault(id(p), {})
                    if 'momentum_buffer' not in param_state:
                        buf = param_state['momentum_buffer'] = Tensor.zeros_like(p)
                    else:
                        buf = param_state['momentum_buffer']
                    
                    buf = momentum * buf + (1 - dampening) * grad
                    
                    if nesterov:
                        grad = grad + momentum * buf
                    else:
                        grad = buf
                    
                    param_state['momentum_buffer'] = buf
                
                # Update parameters
                p.data = p.data - group['lr'] * grad
```

**Adam Optimizer**
```python
class Adam(Optimizer):
    """Adam optimizer with adaptive learning rates."""
    
    def __init__(self, params, lr=1e-3, betas=(0.9, 0.999), eps=1e-8,
                 weight_decay=0, amsgrad=False):
        if not 0.0 <= lr:
            raise ValueError(f"Invalid learning rate: {lr}")
        if not 0.0 <= eps:
            raise ValueError(f"Invalid epsilon value: {eps}")
        if not 0.0 <= betas[0] < 1.0:
            raise ValueError(f"Invalid beta parameter at index 0: {betas[0]}")
        if not 0.0 <= betas[1] < 1.0:
            raise ValueError(f"Invalid beta parameter at index 1: {betas[1]}")
        if not 0.0 <= weight_decay:
            raise ValueError(f"Invalid weight_decay value: {weight_decay}")
        
        defaults = dict(lr=lr, betas=betas, eps=eps,
                       weight_decay=weight_decay, amsgrad=amsgrad)
        super().__init__(params, defaults)
    
    def step(self):
        """Perform a single optimization step."""
        for group in self.param_groups:
            for p in group['params']:
                if p.grad is None:
                    continue
                
                grad = p.grad
                if grad.dtype in {dtypes.float16, dtypes.bfloat16}:
                    grad = grad.float()
                
                amsgrad = group['amsgrad']
                state = self.state.setdefault(id(p), {})
                
                # State initialization
                if len(state) == 0:
                    state['step'] = 0
                    state['exp_avg'] = Tensor.zeros_like(p)
                    state['exp_avg_sq'] = Tensor.zeros_like(p)
                    if amsgrad:
                        state['max_exp_avg_sq'] = Tensor.zeros_like(p)
                
                exp_avg, exp_avg_sq = state['exp_avg'], state['exp_avg_sq']
                if amsgrad:
                    max_exp_avg_sq = state['max_exp_avg_sq']
                beta1, beta2 = group['betas']
                
                state['step'] += 1
                bias_correction1 = 1 - beta1 ** state['step']
                bias_correction2 = 1 - beta2 ** state['step']
                
                # Apply weight decay
                if group['weight_decay'] != 0:
                    grad = grad + group['weight_decay'] * p
                
                # Update biased first moment estimate
                exp_avg = exp_avg * beta1 + grad * (1 - beta1)
                # Update biased second raw moment estimate
                exp_avg_sq = exp_avg_sq * beta2 + (grad * grad) * (1 - beta2)
                
                if amsgrad:
                    # Maintains the maximum of all 2nd moment running avg
                    max_exp_avg_sq = Tensor.maximum(max_exp_avg_sq, exp_avg_sq)
                    # Use the max for normalizing running avg of gradient
                    denom = (max_exp_avg_sq.sqrt() / (bias_correction2 ** 0.5)) + group['eps']
                    state['max_exp_avg_sq'] = max_exp_avg_sq
                else:
                    denom = (exp_avg_sq.sqrt() / (bias_correction2 ** 0.5)) + group['eps']
                
                step_size = group['lr'] / bias_correction1
                
                p.data = p.data - step_size * exp_avg / denom
                
                state['exp_avg'] = exp_avg
                state['exp_avg_sq'] = exp_avg_sq
```

### Learning Rate Schedulers

**Base Scheduler**
```python
class _LRScheduler:
    """Base class for learning rate schedulers."""
    
    def __init__(self, optimizer, last_epoch=-1):
        self.optimizer = optimizer
        if last_epoch == -1:
            for group in optimizer.param_groups:
                group.setdefault('initial_lr', group['lr'])
        else:
            for i, group in enumerate(optimizer.param_groups):
                if 'initial_lr' not in group:
                    raise KeyError(f"param 'initial_lr' is not specified "
                                 f"in param_groups[{i}]")
        self.base_lrs = [group['initial_lr'] for group in optimizer.param_groups]
        self.last_epoch = last_epoch
        self.step()
    
    def get_lr(self):
        raise NotImplementedError
    
    def step(self, epoch=None):
        if epoch is None:
            epoch = self.last_epoch + 1
        self.last_epoch = epoch
        for param_group, lr in zip(self.optimizer.param_groups, self.get_lr()):
            param_group['lr'] = lr
```

## 🧪 Testing Requirements

- [ ] Test SGD with and without momentum
- [ ] Test Adam convergence on simple optimization problems
- [ ] Test parameter grouping with different learning rates
- [ ] Test gradient clipping functionality
- [ ] Test optimizer state saving and loading
- [ ] Test with different parameter shapes and types
- [ ] Test numerical stability with extreme gradients
- [ ] Integration tests with neural network training
- [ ] Performance benchmarks against PyTorch optimizers
- [ ] Test learning rate scheduling

## 📁 Files to Create/Modify

- `grad/optim/__init__.py` - Optimizer package
- `grad/optim/optimizer.py` - Base Optimizer class
- `grad/optim/sgd.py` - SGD implementation
- `grad/optim/adam.py` - Adam implementation
- `grad/optim/rmsprop.py` - RMSprop implementation
- `grad/optim/lr_scheduler.py` - Learning rate schedulers
- `tests/test_optimizers.py` - Comprehensive optimizer tests
- `tests/test_lr_schedulers.py` - Scheduler tests
- `examples/optimizer_comparison.py` - Usage examples

## 🔗 Dependencies

- Requires: Neural network layers (#08) - Parameter class
- Requires: Autograd system (#01, #03) - gradient computation
- Requires: Tensor operations (#06) - mathematical operations
- Requires: Loss functions (#09) - for training examples
- Blocks: Complete training pipeline (#11)
- Blocks: Advanced training features

### Basic Usage
```python
from grad.optim import SGD, Adam
from grad.nn import Sequential, Linear, ReLU

# Create model and optimizer
model = Sequential(Linear(10, 5), ReLU(), Linear(5, 1))
optimizer = SGD(model.parameters(), lr=0.01, momentum=0.9)

# Training step
def train_step(x, y):
    # Forward pass
    output = model(x)
    loss = criterion(output, y)
    
    # Backward pass
    optimizer.zero_grad()
    loss.backward()
    optimizer.step()
    
    return loss

# Adam optimizer
optimizer = Adam(model.parameters(), lr=0.001, betas=(0.9, 0.999))
```

### Parameter Groups
```python
# Different learning rates for different layers
optimizer = SGD([
    {'params': model.layer1.parameters(), 'lr': 0.01},
    {'params': model.layer2.parameters(), 'lr': 0.001}
], momentum=0.9)
```

### Learning Rate Scheduling
```python
from grad.optim.lr_scheduler import StepLR, ExponentialLR

optimizer = Adam(model.parameters(), lr=0.1)
scheduler = StepLR(optimizer, step_size=30, gamma=0.1)

for epoch in range(100):
    train_epoch()
    scheduler.step()
```

## 🚨 Edge Cases to Handle

- Empty parameter lists
- Parameters without gradients
- Very large or small gradients (numerical stability)
- Mixed precision parameters
- Parameters on different devices
- State management for complex optimizers
- Memory leaks in optimizer state
- Thread safety for parallel training

## 🔧 Performance Considerations

- Efficient state management and memory usage
- In-place operations where possible
- Minimize temporary tensor allocations
- Vectorized operations for parameter updates
- Consider future GPU implementation
- Memory pooling for optimizer state tensors

## 🏷️ Labels

- `enhancement`
- `optimizers`
- `training`
- `high-priority`
- `machine-learning`

## 📈 Priority

**High** - Essential for neural network training and parameter optimization

## 🔗 Related Issues

- Depends on: Neural network layers (#08)
- Depends on: Autograd system (#01, #03)
- Depends on: Loss functions (#09)
- Blocks: Complete training pipeline (#11)
- Related to: Performance optimization
- Related to: Advanced training features
